### DISCUSSION TITLE
Client Query Invalidation does not invalidate server side caches

### DISCUSSION THREAD

**Discussion by MokDevelopment** [TanStack/query#4271] [Q&A]
### Describe the bug

When creating a queryClient outside my getServerSideProps (next.js) I am able to cache queries on the server, giving visitors of the app the ability to request a cached api fetch which is not made for every client again and again. 

This works- however when trying to invalidate the cache via 
` queryClient.invalidateQueries()` within

1.  a server side api function I am not able to invalidate the queryClient. Seems as the reference to the queryClient is not the same. I am using new queryCient() instead of useQueryClient as it is not available for for non react components.  
2. Also when trying to invalidate the queryClient on the client (browser) I am able to force a refetch for this particular user / client - but every other new request will get the old version cached by the server.

So how do I invalidate cached queries made by node.js then? Can this be done server side or client side?
Is there a useQueryClient variant for non react components? 

I am not sure if this is a bug or was never planned to be working. 

```
api/invalidate.ts:

import { QueryClient } from 'react-query';

import { NextApiRequest, NextApiResponse } from "next"

const queryClient = new QueryClient();

export default async function handler(
  request: NextApiRequest,
  response: NextApiResponse
) {

  try { 
    // Get QueryClient from the context
    queryClient.invalidateQueries()

    return response.json({status:'ok'})
  }
}
```

```
examplePage.tsx:

const queryClient = new QueryClient()
export async function getServerSideProps(context) {
//preview mode - make an authenticated request
    await queryClient.prefetchQuery(
      [KEY_ARTICLE_LISTING, { id, locale }],
      () => fetchListingWithAuth({ locale, context, path }),
      { staleTime: Infinity }
    )
}

```

 

### Your minimal, reproducible example

not created so far as 1st needs clarification

### Steps to reproduce

1. Create a next.js project
2. Prefetch on the server
3. try to invalidate the cache

### Expected behavior

Caches get invalidated

### Platform

osx 12.1
"react-query": "^3.39.2",
"react": "^18.2.0",
"next": "^12.1.4",

### react-query version

3.39.2

**Comment by TkDodo** [comment]
I don't really know what nextJs does with variables created on the server, outside of `getServerSideProps`. I guess this also depends on how it is deployed - for example, edge functions probably behave differently.

[The docs](https://tanstack.com/query/v4/docs/guides/ssr#on-the-server) state that you should:

> Create a new QueryClient instance inside of your request handler. This ensures that data is not shared between different users and requests.

So I would say support for what you're trying to do is at best undocumented.

There is also a discussion here where the idea was to create the client inside getServerSideProps, and then persist it to redis with the persister plugins:

- https://github.com/TanStack/query/discussions/3563

**Comment by MokDevelopment** [comment]
@TkDodo Thanks for ur input Dominik.
I could not find any solutions regarding this topic. What I found is this stackOverFlow post here:
https://stackoverflow.com/questions/73307822/server-side-api-response-caching-with-react-query-and-next-js

**Reply by TkDodo** [reply]
I guess it also depends with what you mean that `invalidateQueries` doesn't work. Invalidation just marks queries as stale and refetches those that have active observers. Observers are created via `useQuery` so on the server, you have none. So if you'd expect `invalidateQueries` to refetch the queries, this wouldn't happen.

you can use `refetchQueries` though, that always refetches...

**Reply by MokDevelopment** [reply]
What I try to achieve is to keep a stale:infinity cache on the server until the cms changes the content and the server needs to refetch. I expected, that there is some kind of best practice concept for this really pretty common workflow - I like the way ISR in Next.js works... you revalidate caches for a specific page by calling an api route. - This happens automatically via the cms when the content is saved. Really this should be possible in a similar way but on a fetch data level instead of constantly firing api requests with waiting clients vs. delivering old data when dealing with server side react query caches.

Just a question in general ... is there a way to fetch an updated state via react query immediately "after" delivering the cached one? At least then there is no waiting time and just one user gets old data?

Thanks again for clarifying this.

**Reply by TkDodo** [reply]
this would all be easier if you'd show a codesandbox reproduction. I don't know if you're still talking about "on the server", like inside getServerSideProps, or about react-query in general ...

**Reply by MokDevelopment** [reply]
I am talking about ReactQuery in general when used for fetches, that are made "on  the server". 
Background is pretty simple: "We have 10000 of data entities, that change maybe once a week. We would like to keep a cached json request on the server and use it for thousands of users instead of fetching it for every user and page again".

As we cannot define a cache stale time like for example 20 min or 20 hours we would need the ability to invalidate the cache manually instead of by time.

**Reply by slava-lu** [reply]
> I guess it also depends with what you mean that `invalidateQueries` doesn't work. Invalidation just marks queries as stale and refetches those that have active observers. Observers are created via `useQuery` so on the server, you have none. So if you'd expect `invalidateQueries` to refetch the queries, this wouldn't happen.
> 
> you can use `refetchQueries` though, that always refetches...

It would be very useful feature if `queryClient.fetchQuery` sees stale data,  it refetches.

**Reply by TkDodo** [reply]
> It would be very useful feature if queryClient.fetchQuery sees stale data, it refetches.

It does that? You just need to provide a staleTime to `fetchQuery`. This is documented here: https://tanstack.com/query/v4/docs/reference/QueryClient#queryclientfetchquery

**Reply by TkDodo** [reply]
@MokDevelopment the way most people are using fetches on the server is to just fetch per request, in `getServerSideProps`. That will seed the client side cache. This is not ideal but a limitation of nextJs that will likely get better with the new router / layouts rfc.

If you want caching on the server, please see the discussion about persisting the cache somewhere
If you want on the fly invalidation like ISR, you'd need to create an endpoint that you can hit that will then invalidate the cache. pretty much what is documented here: https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#using-on-demand-revalidation

**Reply by slava-lu** [reply]
> > It would be very useful feature if queryClient.fetchQuery sees stale data, it refetches.
> 
> It does that? You just need to provide a staleTime to `fetchQuery`. This is documented here: https://tanstack.com/query/v4/docs/reference/QueryClient#queryclientfetchquery
well, it does not work for me in next.js when putting the `queryClient `outside of the `getServerSideProps`.  It ignores the stale time and does not refetch.

**Reply by TkDodo** [reply]
> well, it does not work for me in next.js when putting the queryClient outside of the getServerSideProps. It ignores the stale time and does not refetch.

show a reproduction?
